File: java/rest-simple/src/test/java/com/example/restsimple/StudentControllerTest.java
package com.example.restsimple;

import com.example.restsimple.controller.StudentController;
import com.example.restsimple.model.Student;
import com.example.restsimple.repository.UpdateStudent;
import com.example.restsimple.service.StudentService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.Collections;
import java.util.Optional;
import java.util.UUID;

import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public class StudentControllerTest {

    private StudentService studentService;
    private MockMvc mockMvc;

    @BeforeEach
    public void setUp() {
        studentService = mock(StudentService.class);
        StudentController studentController = new StudentController(studentService);
        mockMvc = MockMvcBuilders.standaloneSetup(studentController).build();
    }

    @Test
    public void testGetAllStudents() throws Exception {
        Student student = new Student("1", "John", "Doe", null);
        when(studentService.getAllStudents()).thenReturn(Collections.singletonList(student));

        mockMvc.perform(get("/students"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.students[0].id").value("1"))
                .andExpect(jsonPath("$.students[0].name").value("John"))
                .andExpect(jsonPath("$.students[0].lastName").value("Doe"));

        verify(studentService).getAllStudents();
    }

    @Test
    public void testCreateStudent() throws Exception {
        String name = "John";
        String lastName = "Doe";
        Student student = new Student(UUID.randomUUID().toString(), name, lastName, null);
        when(studentService.createStudent(any(Student.class))).thenReturn(student);

        mockMvc.perform(post("/students")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"" + name + "\", \"lastName\":\"" + lastName + "\"}"))
                .andExpect(status().isCreated());

        verify(studentService).createStudent(any(Student.class));
    }

    @Test
    public void testUpdateStudent() throws Exception {
        String id = "1";
        Student student = new Student(id, "John", "Doe", null);
        when(studentService.updateStudent(eq(id), any(UpdateStudent.class))).thenReturn(Optional.of(student));

        mockMvc.perform(put("/students/" + id)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"John\", \"lastName\":\"Smith\"}"))
                .andExpect(status().isOk());

        verify(studentService).updateStudent(eq(id), any(UpdateStudent.class));
    }

    @Test
    public void testDeleteStudent() throws Exception {
        String id = "1";
        Student student = new Student(id, "John", "Doe", null);
        when(studentService.deleteStudent(id)).thenReturn(Optional.of(student));

        mockMvc.perform(delete("/students/" + id))
                .andExpect(status().isNoContent());

        verify(studentService).deleteStudent(id);
    }

    @Test
    public void testGetAllStudents_EmptyList() throws Exception {
        when(studentService.getAllStudents()).thenReturn(Collections.emptyList());

        mockMvc.perform(get("/students"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.students").isEmpty());

        verify(studentService).getAllStudents();
    }

    @Test
    public void testCreateStudent_MissingName() throws Exception {
        mockMvc.perform(post("/students")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"lastName\":\"Doe\"}"))
                .andExpect(status().isBadRequest());

        verify(studentService, never()).createStudent(any(Student.class));
    }

    @Test
    public void testCreateStudent_MissingLastName() throws Exception {
        mockMvc.perform(post("/students")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"John\"}"))
                .andExpect(status().isBadRequest());

        verify(studentService, never()).createStudent(any(Student.class));
    }

    @Test
    public void testUpdateStudent_NotFound() throws Exception {
        String id = "1";
        when(studentService.updateStudent(eq(id), any(UpdateStudent.class))).thenReturn(Optional.empty());

        mockMvc.perform(put("/students/" + id)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"John\", \"lastName\":\"Smith\"}"))
                .andExpect(status().isNotFound());

        verify(studentService).updateStudent(eq(id), any(UpdateStudent.class));
    }

    @Test
    public void testDeleteStudent_NotFound() throws Exception {
        String id = "1";
        when(studentService.deleteStudent(id)).thenReturn(Optional.empty());

        mockMvc.perform(delete("/students/" + id))
                .andExpect(status().isNotFound());

        verify(studentService).deleteStudent(id);
    }
}

File: kotlin/rest-simple/src/test/kotlin/com/example/restsimple/StudentControllerTest.kt
package com.example.restsimple

import com.example.restsimple.model.Student
import com.example.restsimple.repository.UpdateStudent
import com.example.restsimple.request.StudentCreateRequest
import com.example.restsimple.request.StudentUpdateRequest
import com.example.restsimple.service.StudentService
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import java.time.LocalDateTime
import java.util.*

@SpringBootTest
@AutoConfigureMockMvc
internal class StudentControllerTest {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockBean
    private lateinit var studentService: StudentService

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    @Test
    fun `test health endpoint`() {
        mockMvc.perform(get("/healthz"))
                .andExpect(status().isOk)
                .andExpect(content().string("{\"message\":\"OK\"}"))
    }

    @Test
    fun `test get all students`() {
        val student = Student("1", "John", "Doe", LocalDateTime.now())
        whenever(studentService.allStudents).thenReturn(mutableListOf(student))

        mockMvc.perform(get("/students"))
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.students[0].name").value("John"))
    }

    @Test
    fun `test create student`() {
        val studentRequest = StudentCreateRequest("John", "Doe")
        val student = Student("1", "John", "Doe", LocalDateTime.now())
        whenever(studentService.createStudent(anyOrNull())).thenReturn(student)

        mockMvc.perform(
                post("/students")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(studentRequest))
        )
                .andExpect(status().isCreated)
                .andExpect(jsonPath("$.student.name").value("John"))
    }

    @Test
    fun `test successful student update`() {
        val studentId = "1234"
        val studentUpdateRequest = StudentUpdateRequest("James", "Bond")
        val updatedStudent = Student(studentId, "James", "Bond", LocalDateTime.now())

        whenever(studentService.updateStudent(anyOrNull(), anyOrNull()))
                .thenReturn(Optional.of(updatedStudent))

        mockMvc.perform(
                put("/students/$studentId")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(studentUpdateRequest))
        )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.student.id").value(studentId))
                .andExpect(jsonPath("$.student.name").value("James"))
                .andExpect(jsonPath("$.student.lastName").value("Bond"))
    }

    @Test
    fun `test update student not found`() {
        val studentUpdateRequest = StudentUpdateRequest("Jane", "Doe")
        whenever(studentService.updateStudent("1", UpdateStudent(studentUpdateRequest.name, studentUpdateRequest.lastName)))
                .thenReturn(Optional.empty())

        mockMvc.perform(
                put("/students/1")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(studentUpdateRequest))
        )
                .andExpect(status().isNotFound)
    }

    @Test
    fun `test delete student`() {
        val student = Student("1", "John", "Doe", LocalDateTime.now())
        whenever(studentService.deleteStudent("1")).thenReturn(Optional.of(student))

        mockMvc.perform(
                delete("/students/1")
        )
                .andExpect(status().isNoContent)
    }

    @Test
    fun `test delete student not found`() {
        whenever(studentService.deleteStudent("1")).thenReturn(Optional.empty())

        mockMvc.perform(
                delete("/students/1")
        )
                .andExpect(status().isNotFound)
    }
}

